openedx/openedx-platform
# feat: `atlas pull` for XBlock translations | FC-0012

Implements the following proposal:

 - https://github.com/openedx/edx-platform/pull/33166

## Known limitations
 - [x] Test i18n translations:
 - [x] This won't pull other plugin translations such as `edx-val` and any non-XBlock library. A follow up pull request will be created.
 - [x] Address `translation.to_locale(lang_code)` vs `translation.to_locale(translation.get_language())`
 - [x] Fix `get_statici18n_js_url`
 - [x] Refactor settings location and shared variables
 - [x] Consider a dedicated `plugin_translations` edX Platform module independent of XBlock module
 - [x] Fix failed tests
 - [x] Update the method and directory names to match the python implementation in the `0019-oep-58-atlas-translations-design.rst` document
 - [x] Upgrade XBlock to 1.9.0: 
   - [x] https://github.com/openedx/XBlock/pull/691
   - [x] https://github.com/openedx/edx-platform/pull/33948 (done manually via this PR)
 - [x] Ensure non--verbose silent the output is respected in `pull_xblock_translations`, otherwise remove the command

## Testing instructions

Install the Drag and Drop v2 fork which supports OEP-58 translations:

```
devstack $ make cms-shell
edx-platform # pip uninstall xblock-drag-and-drop-v2
edx-platform # pip install -e https://github.com/Zeit-Labs/xblock-drag-and-drop-v2/archive/atlas-translations.zip
edx-platform # make OPENEDX_ATLAS_PULL=yes pull_xblock_translations
devstack $ make cms-restart
```

Then add a Drag and Drop XBlock instance and it should use the translations from the `atlas pull` results instead of the bundled ones. 

Of course, it'll be difficult to distinguish which is which, therefore I've hacked the process as described in the `Testing results`.

## Testing results

I've manually replaced the `atlas pull` French translations with Chinese so I ensure that I'm not just seeing stale translations.

This is a temp. hack during testing and it looks like the Drag and Drop v2 translation is being pulled correctly via atlas.

These are the lines that I've added after the `atlas pull` command:

```py
def atlas_pull_xblock_by_module_names():
    ...
    # This hack purpose is to make it obvious that the translations are working by showing Chinese instead of Arabic.
    for dummy_lang in ['ar', 'fr']:
        subprocess.run(
            args=f'rm -rf {dummy_lang} && cp -r zh_CN {dummy_lang}',
            shell=True,
            cwd=f'{XBLOCK_TRANSLATIONS_ROOT}/drag_and_drop_v2/conf/locale',
            check=True,
        )
```

<details><summary>Screenshots</summary>

**Python code translations**
![Screenshot from 2023-11-12 22-26-02](https://github.com/openedx/edx-platform/assets/645156/cb5d6cd1-c2cd-4632-8199-a3066bfb201c)

**JavaScript code translations**
![Screenshot from 2023-11-12 22-26-20](https://github.com/openedx/edx-platform/assets/645156/2e0616dd-50b7-4795-b332-0e00d3691453)

</details> 

## Refs
This pull request is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which implements the [Translation Infrastructure update OEP-58](https://docs.openedx.org/en/latest/developers/concepts/oep58.html).
